src/gui/dialogs/multiplayer/mp_connect.cpp

```diff
@@ -16,174 +16,304 @@
+#if 0
+void clear_listbox_selection(listbox& listbox)
+{
+	const auto selection = listbox.get_selected_row();
+	if(selection >= 0) {
+		listbox.select_row(selection, false);
+	}
+}
+#endif
+
+std::string clean_textbox_value(text_box& text_box)
+{
+	auto str = text_box.get_value();
+	boost::trim(str);
+	return str;
+}
+
+}
+
+namespace dialogs
+{
+
 /*WIKI
  * @page = GUIWindowDefinitionWML
- * @order = 2_mp_server_list
+ * @order = 2_mp_connect
  *
- * == Multiplayer server list ==
+ * == Multiplayer connect ==
  *
- * This shows the dialog with a list of predefined multiplayer servers.
+ * This shows the dialog to the MP server to connect to.
  *
  * @begin{table}{dialog_widgets}
  *
- * server_list & & listbox & m &
- *         Listbox with the predefined servers to connect to. $
- *
- * -name & & styled_widget & o &
- *         Widget which shows the name of the server. $
+ * start_table & & text_box & m &
+ *         The name of the server to connect to. $
  *
- * -address & & styled_widget & m &
- *         The address/host_name of the server. $
+ * list & & button & o &
+ *         Shows a dialog with a list of predefined servers to connect to. $
  *
  * @end{table}
  */
 
-class mp_server_list : public modal_dialog
+REGISTER_DIALOG(mp_connect)
+
+mp_connect::mp_connect()
+	: host_name_(register_text("host_name",
+							   true,
+							   preferences::network_host,
+							   preferences::set_network_host,
+							   true))
+	, builtin_servers_(preferences::builtin_servers_list())
+	, user_servers_(preferences::user_servers_list())
 {
-public:
-	mp_server_list() : host_name_()
-	{
+	set_restore(true);
+}
+
+void mp_connect::pre_show(window& win)
+{
+	text_box& hostname_box = find_widget<text_box>(&win, "host_name", false);
+	listbox& server_list = find_widget<listbox>(&win, "server_list", false);
+	button& button_add = find_widget<button>(&win, "server_add", false);
+	button& button_del = find_widget<button>(&win, "server_delete", false);
+
+	for(const auto* servers : server_lists()) {
+		for(const auto& server : *servers) {
+			insert_into_server_listbox(server_list, server);
+		}
 	}
 
-	const std::string& host_name() const
-	{
-		return host_name_;
+	select_first_match();
+
+	connect_signal_notify_modified(hostname_box, std::bind(&mp_connect::on_address_change, this));
+	connect_signal_notify_modified(server_list, std::bind(&mp_connect::on_server_select, this));
+	connect_signal_mouse_left_click(button_add, std::bind(&mp_connect::on_server_add, this));
+	connect_signal_mouse_left_click(button_del, std::bind(&mp_connect::on_server_delete, this));
+}
+
+void mp_connect::insert_into_server_listbox(listbox& listbox, const server_info& srv, int pos)
+{
+	const std::map<std::string, string_map>& entry{
+		{ "name",    string_map{{"label", srv.name}} },
+		{ "address", string_map{{"label", srv.address}} },
+	};
+
+	listbox.add_row(entry, pos);
+}
+
+void mp_connect::select_first_match()
+{
+	window* window = get_window();
+	if(!window) {
```

Comment:
You don't need to always check `window` is valid. It's guaranteed to exist as long as the dialog is shown.